### ISSUE TITLE
Proposal: Installer for Windows

### ISSUE THREAD

**Issue by Bill-Stewart** [syncthing/syncthing#8082]
I wrote an installer package for Windows (built using Inno Setup) that can install Syncthing for all users (Windows service using NSSM) or for the current user (automatic start at logon). It includes some helper scripts that simplify things for users on Windows and should hopefully reduce the number of questions from Windows users. Documentation here:

https://gist.github.com/Bill-Stewart/67fe1ad44bbc71c6b3d060f0fef3302a

Feedback is welcome.

Syncthing maintainers: If you want to incorporate this installer into the build process on Windows, feel free to do so.

**Comment by Bill-Stewart** [comment]
Initial commit has been published to Github:

https://github.com/Bill-Stewart/SyncthingWindowsSetup

Do I need permission to publish the installer executable?

**Comment by imsodin** [comment]
No need for any permissions, you are free to publish your own work. And as long as you are not doing anything obviously nefarious or against the Syncthing principles, or incorrectly trying to appear "official", there's no issue with using the name "syncthing" in your thing either. It's cool if there's more ways to make use of Syncthing. You could also add your project here: https://docs.syncthing.net/users/contrib.html

I am not so sure about integrating it: The utilities for the firewall and auto-choosing 32/64bit seem nice (among others), however we do recommend SyncTrayzor on windows which isn't part of the installer. And there's a lot of different ways to run Syncthing automatically on windows with various respective advantages (documented (partly?) on https://docs.syncthing.net/users/autostart.html#windows). I don't know enough about these to justify making one of those the default.

**Comment by Bill-Stewart** [comment]
As noted at https://docs.syncthing.net/users/autostart.html#windows: "There is currently no official installer available for Windows." This installer could fill that role -- it covers the scenarios listed there (and also provides some necessary things not mentioned, such as adding a firewall rule automatically). As I mentioned - the maintainers are free to use this if they feel so inclined. Cheers!

**Comment by calmh** [comment]
I think the path to officialdom lies via widespread use. Initially it should be listed on the contributions page, and we could add a sentence to the download page saying "there is a windows installer here" as well.

**Comment by Bill-Stewart** [comment]
The latest installer update takes advantage of the v1.19.0 `--skip-port-probing` option in the configuration script (thanks for adding that!).

**Comment by Bill-Stewart** [comment]
Update: In an effort to reduce the irritatingly high number of malware false-positives with various security software, [Syncthing Windows Setup](https://github.com/Bill-Stewart/SyncthingWindowsSetup/) has switched to running the service using [shawl](https://github.com/mtkennerly/shawl) rather than NSSM.

**Comment by Bill-Stewart** [comment]
Another update: Due to the prevalence of "PowerShell paranoia" (a term I just made up), the next version of Syncthing Windows Setup will use a separate standalone executable to create/update the local service account and Windows service (if installing for all users/as administrator).

**Comment by Bill-Stewart** [comment]
Yet another update: It seems that some anti-malware configurations block process creation using WMI, which is exactly what the "Start Syncthing" and "Stop Syncthing" helper scripts were using to run `syncthing.exe` using lower priority in a hidden window. (Aside: Starting a process under the context of a particular user means the process runs _as that user_ and doesn't somehow gain magical powers just because WMI launched it.) Nevertheless, due to this "WMI process creations are always bad" paranoia, the next version of Syncthing Windows Setup will use yet another standalone executable for starting and stopping Syncthing as the current user. (Silly, but it is what it is.)

**Comment by calmh** [comment]
Is this all because Syncthing doesn't support natively running as a service, or is it because of the CLI nature of the binary (that it pops a console window)? Supporting running directly as a service (which by definition hides the window, I think) isn't particularly tricky, I think.

**Comment by bt90** [comment]
https://pkg.go.dev/golang.org/x/sys/windows/svc

Seems to provide the necessary bits, including a function to detect if the program is started as a service or otherwise.

**Comment by Bill-Stewart** [comment]
Let me see if I can answer...

1. Syncthing Windows Setup supports admin install (per-computer), which will install a service. In this case, we don't care about the console window because it won't be visible to the user anyway.

2. Syncthing Windows Setup also supports non admin install (per-user). In this case, we just launch `syncthing.exe`. The nuisance is that in this case the console window persists. The workaround for this is to start it with `SW_HIDE` window state (hidden). That's what the script was doing using WMI and what the new tool now does using the `CreateProcess` Windows API (i.e., it starts Syncthing in a hidden window).

There are pros and cons to both install modes:

1. If you install per-computer (service), then Syncthing Windows Setup creates a local user account whose sole purpose is to run the service. This is for security reasons: If the service were to be compromised, the access would be limited only to this service account, which is not privileged and has no group memberships granting it access to anything (except for the folders that have to be synced, which is unavoidable). The complication in this scenario is that the user has to manually grant the service account access to the synced folders. The advantage, of course, is that folders sync even when no users are logged on.

2. If you install per-user, then Syncthing simply runs as the current user. There's no service account in this scenario and normally no need to add permissions to folders owned by the user. The downside is that you have to log on before Syncthing runs.

I would say that running as a service has its advantages, but there are security implications (which is why Syncthing Windows Setup creates a separate local user account). Running as a current user works too, but in that case there's the console window to deal with. Syncthing Windows Setup deals with both scenarios and the user can choose which they want.

Does that answer it?

**Comment by calmh** [comment]
I think the answer I was looking for then is "no, just being able to run as a service natively wouldn't help massively".

**Comment by Bill-Stewart** [comment]
If the primary reason for adding service capability is to avoid the console window, then IMO the LOE involved isn't worth it, particularly considering the security implications and added support overhead. In any case, Syncthing Windows Setup already does it (using [shawl](https://github.com/mtkennerly/shawl)).

**Comment by calmh** [comment]
Nah, I was more thinking that adding service capability is easy and if all your contortions were due to that lack we could quite easily fix it.

**Comment by Bill-Stewart** [comment]
Yes, it would be nice if `syncthing.exe` had a Windows-specific service installation option. Certainly not critical, but "nice to have." If this was available there wouldn't be a need for shawl/nssm/etc.

**Comment by necr0x** [comment]
Было бы здорово, если бы пользователи Windows перешли на использование Linux.

**Comment by lld1995** [comment]
--no-console doesn't work on win11

**Comment by rasa** [comment]
> --no-console doesn't work on win11

@lld1995 I just tested it on Windows 11, and it worked as expected. The window was hidden. And I then unhide the window using `cmdow`.

**Comment by tomasz1986** [comment]
@rasa Hiding doesn't work when using Windows Terminal instead of the old Command Prompt. In Windows 11, contrary to previous versions of the OS, Terminal is set as the default console.

**Comment by rasa** [comment]
> [@rasa](https://github.com/rasa) Hiding doesn't work when using Windows Terminal instead of the old Command Prompt. In Windows 11, contrary to previous versions of the OS, Terminal is set as the default console.

@tomasz1986 Thanks for the clarification. I'm assuming 
```
start "syncthing" cmd /c syncthing --no-console
```
will work even instead a wt window, yes?

**Comment by MadSpindel** [comment]
> Nah, I was more thinking that adding service capability is easy and if all your contortions were due to that lack we could quite easily fix it.

Installer is not needed for Windows. It's possible to install Syncthing on Windows with winget today:
```
winget install Syncthing.Syncthing
```

It uses the following:
```
Installer:
  Installer Type: portable (zip)
  Installer Url: https://github.com/syncthing/syncthing/releases/download/v2.0.13/syncthing-windows-amd64-v2.0.13.zip
  Installer SHA256: eeb1ff8afb0bbbc1fbd0f640b0fd7c00cdf211b82d5be52ea30a5d76a34f6635
  Release Date: 2026-01-06
  Offline Distribution Supported: true
```
it creates a symlink in WinGet/Links to syncthing.exe:
```
~> which syncthing
╭───┬───────────┬───────────────────────────────────────────────────────────────────┬──────────╮
│ # │  command  │                               path                                │   type   │
├───┼───────────┼───────────────────────────────────────────────────────────────────┼──────────┤
│ 0 │ syncthing │ C:\Users\fredr\AppData\Local\Microsoft\WinGet\Links\syncthing.exe │ external │
╰───┴───────────┴───────────────────────────────────────────────────────────────────┴──────────╯
```

It would be nice if syncthing.exe could speak with Windows Service Control Manager (like report its status). Then to add it as a service on Windows would be something like:
```
sc.exe create "Syncthing Service" binpath= "C:\Users\fredr\AppData\Local\Microsoft\WinGet\Links\syncthing.exe --no-console --no-browser"
```

And the install instructions would be very clear, just use Winget. And if you want to autostart, use the sc.exe command as above. As @bt90 wrote, I think this is needed: https://pkg.go.dev/golang.org/x/sys/windows/svc

**Comment by Bill-Stewart** [comment]
> Then to add it as a service on Windows would be something like...

It's not _quite_ that simple. Whatever user account starts the service needs to have `SeServiceLogonRight` and the password needs to be managed somehow (these tasks generally require administrator access). These complexities have been taken into account in the Syncthing Windows Setup installer. (For most users, it's recommended not to install the service and just simply install for the current user.)

**Comment by MadSpindel** [comment]
> > Then to add it as a service on Windows would be something like...
> 
> It's not _quite_ that simple. Whatever user account starts the service needs to have `SeServiceLogonRight` and the password needs to be managed somehow (these tasks generally require administrator access). These complexities have been taken into account in the Syncthing Windows Setup installer. (For most users, it's recommended not to install the service and just simply install for the current user.)

Maybe I should've wrote "sudo" before, because I was suggesting to just use Local system account. I guess my main point is that is more important to make it compatible with Service Control Manager anyway. The rest the user can google or use your installer.

**Comment by Bill-Stewart** [comment]
> I was suggesting to just use Local system account

This would be akin to running Syncthing as root on a Linux system (and not a good idea for the same reasons). Syncthing Windows Setup doesn't even run the service using the built-in `LocalService` account (SID S-1-5-19) because `LocalService` is a member of `Users` by default, which (IMO) is still too much access. (Syncthing Windows Setup actually creates a local user account to run the service, for users that actually want to run the service.)

**Comment by MadSpindel** [comment]
> > I was suggesting to just use Local system account
> 
> This would be akin to running Syncthing as root on a Linux system (and not a good idea for the same reasons). Syncthing Windows Setup doesn't even run the service using the built-in `LocalService` account (SID S-1-5-19) because `LocalService` is a member of `Users` by default, which (IMO) is still too much access. (Syncthing Windows Setup actually creates a local user account to run the service, for users that actually want to run the service.)

If I want it to run before I've logged in I don't see what's wrong with it? My computer, my choice. I see many different services with Local System when I check my Services (like Tailscale, Chrome Remote Desktop etc.). And speaking about security I would never install a program with a Third-party installer.

**Comment by Bill-Stewart** [comment]
> If I want it to run before I've logged in I don't see what's wrong with it? My computer, my choice.

Of course; you're free to disregard whatever security recommendations you like if you own the computer.

> I would never install a program with a Third-party installer.

That's also up to you, of course. The installer I wrote, and all of its helper tools, are open-source, and are designed to provide a seamless (and as secure as possible) installation experience for Windows users.
